### PR TITLE
As I user, I can play a human vs computer, computer vs human, and computer vs computer game

### PR DESCRIPTION
**Pull request by ellehallal**
- Board: initialise with default squares. Amend available squares to return available squares
- Player: Initialise with `mark`, `name`, and `display`.
- ComputerPlayer & HumanPlayer: create class with `choose_move` method.
- PlayerMaker: create class to initialise different player classes
- PlayerSelector: create class. Validates user input before initialising a new player type class
- GameFactory: create class. Creates new instance of game with `player selector`      
- Controller: Initialise with `display` and `game factory`, add `game_setup`
- Display: remove user input prompt, amend `show_current_player`. Add `Ask_for_player`, `show_computer_thinking`, `show_computer_move`
- Game: split `play_move` and `toggle_current_player`. `Remove reset_game`
- GameManager: remove `@game`,  remove `game_selection`
- To do: refactor and change class names to be more descriptive

### CODE REVIEW

File: lib/board.rb [hunk 1]
```diff
@@ -36,14 +36,22 @@ def winning_player_exists?(player1_mark, player2_mark)
...
-    available_squares.zero?
+    available_squares_quantity.zero?
   end
 
   def clear_squares
     @squares = [1, 2, 3, 4, 5, 6, 7, 8, 9]
   end
 
-  def available_squares
+  def available_squares_quantity
```
Comment (matthewglover) [inline lib/board.rb:46]: Maybe `total_available_squares`?

File: lib/board.rb [hunk 2]
```diff
@@ -36,14 +36,22 @@ def winning_player_exists?(player1_mark, player2_mark)
...
     @squares.count { |square| square.is_a? Integer }
   end
+
+  def available_squares
+    available = []
+    @squares.each do |square|
```
Comment (matthewglover) [inline lib/board.rb:52]: Could you use `select`?

File: lib/board.rb [hunk 3]
```diff
@@ -36,14 +36,22 @@ def winning_player_exists?(player1_mark, player2_mark)
...
+    available_squares_quantity.zero?
   end
 
   def clear_squares
     @squares = [1, 2, 3, 4, 5, 6, 7, 8, 9]
   end
 
-  def available_squares
+  def available_squares_quantity
     @squares.count { |square| square.is_a? Integer }
```
Comment (matthewglover) [inline lib/board.rb:47]: There is duplication here - ideally this should use the `available_squares` method and then just call `size`

File: lib/controller.rb [hunk 4]
```diff
@@ -1,23 +1,28 @@
+require_relative 'game_factory.rb'
+
 class Controller
-  attr_reader :game
-  def initialize(game, display)
-    @game = game
+  attr_reader :new_game
```
Comment (matthewglover) [inline lib/controller.rb:4]: Is this used?

File: lib/controller.rb [hunk 5]
```diff
@@ -1,23 +1,28 @@
...
+
+  def initialize(display, game_factory)
     @display = display
+    @game_factory = game_factory
   end
 
-  def main_game
-    @game.over?
+  def main_game(squares = [1, 2, 3, 4, 5, 6, 7, 8, 9])
+    game_setup(squares)
```
Comment (matthewglover) [inline lib/controller.rb:12]: Could you just pass in a game - as squares are immediately used to create a game.

File: lib/game.rb [hunk 6]
```diff
@@ -28,9 +20,17 @@ def winning_player
...
 
-  def play_move(move)
-    @board.player_make_move(@current_player.mark, move.to_i)
-    toggle_current_player
+  def play_move
+    player_move = @current_player.choose_move(@board)
+    @board.player_make_move(@current_player.mark, player_move)
+  end
+
+  def toggle_current_player
```
Comment (matthewglover) [inline lib/game.rb:28]: Could be private?

File: lib/game_manager.rb [hunk 7]
```diff
@@ -11,26 +10,18 @@ def play
   end
 
   def play_again?
```
Comment (matthewglover) [inline lib/game_manager.rb:12]: Can be private?

File: lib/board.rb (hunk 1 above)
Comment (ellehallal) [reply lib/board.rb:46]: `total_available_squares` is a better name for this method. Thanks, Matt

File: lib/board.rb (hunk 2 above)
Comment (ellehallal) [reply lib/board.rb:52]: Ah, yes! Thank you

File: lib/controller.rb (hunk 4 above)
Comment (ellehallal) [reply lib/controller.rb:4]: It isn't. I'll remove it, thanks

File: lib/game.rb (hunk 6 above)
Comment (ellehallal) [reply lib/game.rb:28]: Yes, I agree

File: lib/game_manager.rb (hunk 7 above)
Comment (ellehallal) [reply lib/game_manager.rb:12]: It can indeed. Thanks

File: lib/controller.rb (hunk 5 above)
Comment (ellehallal) [reply lib/controller.rb:12]: Good point. I'll change this

**Pull request merged**